src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -125,24 +157,66 @@ bool DeepChangeChecker::check_outgoing_links(TableKey table_key, Table const& ta
     };
 
     const Obj obj = table.get_object(ObjKey(obj_key));
-    auto linked_object_changed = [&](OutgoingLink const& link) {
-        if (already_checking(link.col_key))
+    auto linked_object_changed = [&](ColKey const& outgoing_link_column) {
+        if (already_checking(outgoing_link_column))
             return false;
-        if (!link.is_list) {
-            if (obj.is_null(ColKey(link.col_key)))
+        if (!outgoing_link_column.is_collection()) {
+            if (obj.is_null(outgoing_link_column))
                 return false;
-            auto dst = obj.get<ObjKey>(ColKey(link.col_key)).value;
-            return check_row(*table.get_link_target(ColKey(link.col_key)), dst, depth + 1);
+            ObjKey dst = obj.get<ObjKey>(outgoing_link_column);
+            REALM_ASSERT(dst);
+            return check_row(*table.get_link_target(outgoing_link_column), dst.value, depth + 1);
         }
 
-        auto& target = *table.get_link_target(ColKey(link.col_key));
-        auto lvr = obj.get_linklist(ColKey(link.col_key));
-        return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
-            return this->check_row(target, key.value, depth + 1);
-        });
+        if (outgoing_link_column.is_list()) {
+            auto& target = *table.get_link_target(outgoing_link_column);
+            auto lvr = obj.get_linklist(outgoing_link_column);
+            return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
+                return this->check_row(target, key.value, depth + 1);
+            });
+        }
+        REALM_ASSERT_EX(outgoing_link_column.is_dictionary() || outgoing_link_column.is_set(),
+                        outgoing_link_column.get_type());
+
+        auto collection_ptr = obj.get_collection_ptr(outgoing_link_column);
+        REALM_ASSERT(collection_ptr);
+        const size_t coll_size = collection_ptr->size();
+        if (outgoing_link_column.get_type() == col_type_Link) {
+            TableRef linked_table = collection_ptr->get_target_table();
+            REALM_ASSERT(linked_table);
+            for (size_t i = 0; i < coll_size; ++i) {
+                auto val = collection_ptr->get_any(i);
+                if (val.is_type(type_Link)) {
+                    auto obj_link = val.get<ObjKey>();
+                    REALM_ASSERT(obj_link);
+                    if (this->check_row(*linked_table, obj_link.value, depth + 1)) {
```

Comment:
The explicit call through `this` in the previous version of this code was due to it being in a lambda that needed to capture `this` by value and isn't applicable in the new code.